File: backend/app/schemas/ai_assistant.py
from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str


class ChatProgress(BaseModel):
    asked: int
    estimated_total: int


class ProposedGoal(BaseModel):
    name: str
    goal_type: Literal[
        "savings", "debt_payoff", "emergency_fund", "custom"
    ] = "savings"
    # Nullable so Claude can omit a number when uncertain; filtered out at
    # apply time if missing or non-positive.
    target_amount: float | None = None
    target_date: date | None = None
    planned_monthly_contribution: float | None = None
    color: str | None = None

    @field_validator("target_amount")
    @classmethod
    def _non_negative_target(cls, v: float | None) -> float | None:
        if v is not None and v < 0:
            raise ValueError("target_amount must be non-negative")
        return v

    @field_validator("planned_monthly_contribution")
    @classmethod
    def _positive_contribution(cls, v: float | None) -> float | None:
        if v is not None and v < 0:
            raise ValueError("planned_monthly_contribution must be non-negative")
        return v


class ProposedBudgetAllocation(BaseModel):
    category_name: str
    amount: float
    is_locked: bool = False

    @field_validator("amount")
    @classmethod
    def _positive_amount(cls, v: float) -> float:
        if v < 0:
            raise ValueError(
                "Budget allocations must be positive — Kallio stores expenses "
                "as positive budgeted amounts."
            )
        return v


class ProposedGoalContribution(BaseModel):
    goal_name: str
    monthly_amount: float

    @field_validator("monthly_amount")
    @classmethod
    def _positive_monthly(cls, v: float) -> float:
        if v < 0:
            raise ValueError("monthly_amount must be positive")
        return v


class ProposedBudget(BaseModel):
    monthly_income: float
    period_type: str = "monthly"
    allocations: list[ProposedBudgetAllocation] = Field(default_factory=list)
    goal_contributions: list[ProposedGoalContribution] = Field(default_factory=list)


class AiProposal(BaseModel):
    monthly_income: float | None = None
    categorization: dict[str, Any] = Field(default_factory=dict)
    goals: list[ProposedGoal] = Field(default_factory=list)
    budget: ProposedBudget | None = None


class ChatRequestContext(BaseModel):
    include_transactions: bool = True


class ChatRequest(BaseModel):
    messages: list[ChatMessage]
    context: ChatRequestContext = Field(default_factory=ChatRequestContext)


class ChatResponse(BaseModel):
    phase: Literal["questioning", "proposing"]
    message: str
    progress: ChatProgress | None = None
    proposal: AiProposal | None = None


class ApplyAiProposalRequest(BaseModel):
    proposal: AiProposal


class ApplyAiProposalResponse(BaseModel):
    categories_created: int
    goals_created: int
    budget_allocations_saved: int
    transactions_categorized: int
    snapshot_id: str | None = None

File: frontend/src/api/aiAssistant.ts
import { apiFetch } from "@/api/client";

export interface AiChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface AiChatProgress {
  asked: number;
  estimated_total: number;
}

export interface AiProposedGoal {
  name: string;
  goal_type: "savings" | "debt_payoff" | "emergency_fund" | "custom";
  target_amount: number;
  target_date: string | null;
  planned_monthly_contribution: number | null;
  color: string | null;
}

export interface AiProposedAllocation {
  category_name: string;
  amount: number;
  is_locked: boolean;
}

export interface AiProposedGoalContribution {
  goal_name: string;
  monthly_amount: number;
}

export interface AiProposedBudget {
  monthly_income: number;
  period_type: string;
  allocations: AiProposedAllocation[];
  goal_contributions: AiProposedGoalContribution[];
}

export interface AiProposal {
  monthly_income: number | null;
  categorization: Record<string, unknown>;
  goals: AiProposedGoal[];
  budget: AiProposedBudget | null;
}

export interface AiChatResponse {
  phase: "questioning" | "proposing";
  message: string;
  progress: AiChatProgress | null;
  proposal: AiProposal | null;
}

export interface AiApplyResponse {
  categories_created: number;
  goals_created: number;
  budget_allocations_saved: number;
  transactions_categorized: number;
  snapshot_id: string | null;
}

export async function aiAssistantChat(
  messages: AiChatMessage[],
  token: string
): Promise<AiChatResponse> {
  return apiFetch<AiChatResponse>(
    "/api/v1/ai/assistant/chat",
    {
      method: "POST",
      body: JSON.stringify({
        messages,
        context: { include_transactions: true },
      }),
    },
    token
  );
}

export async function aiAssistantApply(
  proposal: AiProposal,
  token: string
): Promise<AiApplyResponse> {
  return apiFetch<AiApplyResponse>(
    "/api/v1/ai/assistant/apply",
    {
      method: "POST",
      body: JSON.stringify({ proposal }),
    },
    token
  );
}
